lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx

```diff
@@ -23,6 +34,69 @@ export type DashboardActivityFiltersProps = {
   onClearSelection?: () => void;
 };
 
+function usePaginatedAPIFetch<
+  Prop extends string,
+  ListType extends any[],
+  FetchType extends { [P in Prop]: ListType } & { pagination: Pagination },
+>(
+  prop: Prop,
+  path: string,
+  filters: Record<string, string | string[] | undefined>,
+): [ListType | undefined, FetchResult<FetchType>, () => void] {
+  const [urlFromPagination, setURLFromPagination] = useState<string>();
+  const [currentList, setCurrentList] = useState<ListType>();
+
+  const fetchResult = useAPIFetch<FetchType>(
+    urlFromPagination ?? path,
+    !urlFromPagination ? filters : undefined,
+  );
+
+  const loadNextPage = useCallback(() => {
+    if (!fetchResult.isLoading && fetchResult.data?.pagination.next) {
+      setURLFromPagination(fetchResult.data.pagination.next);
+    }
+  }, [fetchResult.data?.pagination.next, fetchResult.isLoading]);
+
+  // Every time data is loaded, append the result to the list
+  useEffect(() => {
+    setCurrentList(prev => {
+      if (!fetchResult.data) {
+        return prev;
+      }
+
+      return [...(prev ?? []), ...fetchResult.data[prop]] as ListType;
+    });
+  }, [prop, fetchResult.data]);
+
+  // Every time filters change, discard previous pages and start from scratch
+  const prevFiltersRef = useRef<typeof filters>(filters);
+  useEffect(() => {
+    const prevFilters = JSON.stringify(prevFiltersRef.current);
+    const newFilters = JSON.stringify(filters);
+
+    if (prevFilters !== newFilters) {
+      setCurrentList(undefined);
+      setURLFromPagination(undefined);
+    }
+
+    prevFiltersRef.current = filters;
+  }, [filters]);
+
+  return [currentList, fetchResult, loadNextPage];
```

Comment:
I think we should be able to simplify this return with a single object tracking the list, loading state and the `loadNextPage` callback, but maybe we can do that change if we decide to extract this hook and reuse it in other places.